### DISCUSSION TITLE
How to remove upload image from storage(s3 or fake-s3 or minio..etc.)

### DISCUSSION THREAD

**Discussion by warnus** [outline/outline#2925] [Self Hosting]
Hello,

Thanks for the good project.

I'm curios how to remove upload image from storage. 

There is no remove image button.

If I just delete Image Node that just node is gone but image file is still exist in storage.
Is there any way to remove image from storage?

I want develop file upload feature by myself. I'm developing now. But, I don't know if there is feature that remove file in storage or not.

Sorry, I'm not familiar with prosemirror. It is hard to programming with that to me.

**Comment by tommoor** [comment]
Images are kept in storage until the document is deleted as they exist in the documents history. When the document is deleted permanently image files are removed at this time.

**Reply by MarioNoll** [reply]
I would have expected that this is also the case for images that have been removed via the delete icon but it seems that is not the case. 

Create Document -> Add Image -> Delete Document -> Permanently Delete = Image gets removed from storage
Create Document -> Add Image -> Delete Image -> Delete Document -> Permanently Delete = Image remains in storage

**Reply by tommoor** [reply]
Yes, there are ways that the images can slip through still. I'm not sure it will ever be possible to delete every file in every situation given the loose link between documents<>attachments. 

This idea seems like a good way around: https://github.com/outline/outline/discussions/3792 – some UI for seeing and managing file attachments in the storage

**Reply by FranciscoKnebel** [reply]
Ran into this while migrating data from a previous platform to Outline. When testing scripts for importing, I created a ton of attachments that weren't attached to documents. Testing how the API works created a lot of dangling attachments that are currently in storage but unreferenced by any document.

Couldn't find any clear way through the API to find these files with no references, delete the attachments and remove the files from storage.

**Reply by steffen-kdab** [reply]
FWIW I ran into the same when developing a migration script, 30GB of "stuff" left over in the uploads storage after deleting all collections. But remember to empty the Trash - it took a while, but it cleaned up all attachments in storage :+1:
